### PR TITLE
Replace `confirmCloseAllTabs` with `confirmCloseOn`

### LINKED ISSUE #5301: closeTab should present a confirmation dialog
# Description of the new feature/enhancement
`closeTab` should present a confirmation dialog if multiple panes are open. This would prevent users from accidentally closing all panes in a tab by accident.

### PR DESCRIPTION
**Pull request by zadjii-msft**
This PR adds support for a more nuanced enum-backed version of `confirmCloseAllTabs`. 

The old setting was simply a bool true/false, and when set to true, it would warn the user when closing a terminal with more than one pane in it. This lacked fidelity for a variety of scenarios that I enumerated in https://github.com/microsoft/terminal/issues/6549#issuecomment-675494526.

So instead, this PR changes that setting to a flag-based enum value. We're adding `warning.confirmCloseOn`, which supports a number of values, which can be combined with one another:

* `"multiplePanes"` warn if closing any tab with more than one pane
* `"multipleTabs"` warn if closing a window with more than one tab
* **NOT IMPLEMENTED IN THIS PR**: `"multipleProcesses"` warn if closing a pane with more than one client attached
* `"always"` warn any time a pane/tab/window is about to be closed. Even if there is one pane in one tab with one client, warn. 
  - notably, this is different than `all` (if all === `multiplePanes|multipleTabs|multipleProcesses`)! That full set doesn't warn on one process in one pane.
* `"never"` I think you can figure this out

The full behavior of these settings are captured in this spreadsheet:

image

**Settings JSON Before / After**

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td>

```jsonc
{
  // Legacy boolean (default: true)
  "warning.confirmCloseAllTabs": true
}
```

</td>
<td>

```jsonc
{
  // Flags enum — combine any of:
  // "always", "multipleTabs",
  // "multiplePanes", "multipleProcesses"
  "warning.confirmCloseOn": ["multipleTabs", "multiplePanes"]
}
```

</td>
</tr>
<tr>
<td>

```jsonc
{
  "warning.confirmCloseAllTabs": false
}
```

</td>
<td>

```jsonc
{
  // Never warn
  "warning.confirmCloseOn": []
}
```

</td>
</tr>
<tr>
<td><em>N/A — not possible before</em></td>
<td>

```jsonc
{
  // Always warn on every close action
  "warning.confirmCloseOn": ["always"]
}
```

</td>
</tr>
<tr>
<td><em>N/A — not possible before</em></td>
<td>

```jsonc
{
  // Only warn when closing a tab with
  // multiple panes (not on window close)
  "warning.confirmCloseOn": ["multiplePanes"]
}
```

</td>
</tr>
</table>

Looping back, these were the originally spec'd scenarios I had called out, and how those settings work after this PR:

| # | Scenario | Setting |
|---|----------|---------|
| 1 | Never be prompted for confirmation | `"warning.confirmCloseOn": []` or `"warning.confirmCloseOn": ["never"]`  |
| 2 | Prompted when closing window with multiple tabs | `"warning.confirmCloseOn": ["multipleTabs"]` |
| 3 | Prompted when closing a tab with multiple panes (#5301) | `"warning.confirmCloseOn": ["multiplePanes"]` |
| 4 | Prompted when closing a pane with multiple clients[^1] (#6549)(#5065)(#10784)(countless other dupes) | `"warning.confirmCloseOn": ["multipleProcesses"]` |
| 5 | Prompted when closing a pane, always | `"warning.confirmCloseOn": ["always"]` |
| 6 | (possibly more scenarios) | (other ones haven't come up yet in the last 5 years) |
| 7 | Prompted only for multiple clients, not multiple tabs/panes[^1] | `"warning.confirmCloseOn": ["multipleProcesses"]` |
| 8 | Prompted always, even one process / one pane / one tab (#12827) | `"warning.confirmCloseOn": ["always"]` |

[^1]: `multipleProcesses` isn't implemented yet. Leonard is working on in-proc conpty, which will make that implementation trivial to do correctly. Without that, it is hard to correctly determine the full set of clients attached to a particular console server (without adding additional IPC to conpty)

* closes #5301
* references #6549
  - we can't close this till we add `multipleProcesses`
* references #5065
* references #10784
* references #12827

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/InteractionViewModel.cpp [hunk 1]
```diff
@@ -18,4 +18,68 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
+        WI_UpdateFlag(current, Model::ConfirmCloseOn::MultipleTabs, winrt::unbox_value<bool>(on));
+        ConfirmCloseOn(current);
+    }
+
+    void InteractionViewModel::SetConfirmCloseOnMultiplePanes(winrt::Windows::Foundation::IReference<bool> on)
+    {
+        auto current = ConfirmCloseOn();
+        WI_UpdateFlag(current, Model::ConfirmCloseOn::MultiplePanes, winrt::unbox_value<bool>(on));
+        ConfirmCloseOn(current);
+    }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/InteractionViewModel.cpp]: The updated preview doesn't get displayed when you call any of these.

Easy fix though, just add this:
`_NotifyChanges(L"ConfirmCloseOnPreview")`

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 2]
```diff
@@ -782,6 +801,28 @@ namespace winrt::TerminalApp::implementation
...
+                    {
+                        if (warningResult != ContentDialogResult::Primary)
+                        {
+                            co_return;
+                        }
+                    }
+                    else
+                    {
+                        co_return;
+                    }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TabManagement.cpp]: nit: I kinda liked the way you did it earlier in this file more

```suggestion
                    const auto strong = weak.get();
                    if (!string || warningResult != ContentDialogResult::Primary)
                    {
                        co_return;
                    }
```

**Comment by dsx724** [comment]
Can this get reviewed?

**Comment by xenago** [comment]
Author is away temporarily on leave, bumping so this issue is kept open

**Comment by DHowett** [comment]
@xenago are you one of zadjii-msft's friends? As his superior, I am surprised to find out that he is on leave. Do you have more information?

**Comment by xenago** [comment]
The author's github profile previously said he was, but it looks like it's just been updated to remove that message!

**Comment by zadjii-msft** [comment]
I'm gonna hand this off officially to @carlos-zamora, since it's no longer hackathon and I can't really iterate on this PR.

--------------

> Long-term, I think the golden-path should be to have a "don't ask me again" checkbox on the content dialogs (referenced as a checkbox in "Warnings" section https://github.com/microsoft/terminal/issues/10000). It's more discoverable and it doesn't require users to go into the settings.

moving discussion from teams:

I don't _love_ the idea of making this an in-dialog checkbox. If we can also have similar checkboxes in the settings UI though, I suppose it is fine. But then we're storing a user preference in `state.json`, when I could very much see this being something that the user wants to move from machine to machine. Not like `settings.json` is totally portable, but I bet folks move that and not `state.json`

Also the proposed migration of the setting into the state.json would mean that users that _currently_ have `warning.confirmCloseAllTabs: false` would migrate to `warning.alwaysConfirmClose: false` _without accompanying state_, so users will get re-bothered by that dialog. Minor, sure, but it is something to be aware of.

**Comment by DHowett** [comment]
> I don't _love_ the idea of making this an in-dialog checkbox.

On the contrary, I think that every dialog and infobar we display should be instantly permanently dismissible.

This one has some nuance though. Don't ask me again, don't ask me again for just one process, don't ask me again for multiple processes, don't ...

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 3]
```diff
@@ -808,15 +808,8 @@ namespace winrt::TerminalApp::implementation
...
-                        {
-                            co_return;
-                        }
-                    }
-                    else
+                    const auto strong = weak.get();
+                    if (!strong ||  warningResult != ContentDialogResult::Primary)
                     {
                         co_return;
                     }
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TabManagement.cpp]: why are we bothering to make sure the object exists at all? we don't USE it after this point.

**Comment by carlos-zamora** [comment]
Made some BIG changes here. I'll update the PR body tomorrow to be more descriptive, but here's some things that stand out:
- changed `confirmCloseOn` to an enum with `never`, `automatic`, `always`
   - `automatic` presents a popup when multiple panes/tabs are about to be closed. 
   - `automatic` be expanded in the future to consider things like "displaying progress ring", "seems to be using shell integration marks and the last mark indicates something got launched", and "multiple processes in one terminal"
- there's a "don't ask me again" checkbox. When checked, we update the setting to `never`.
   - This _does_ cause a settings reload, which can be annoying if it resets the font size or something like that. That said, that's a known bug and I'm planning on working on that soon for auto-save. Sounds like I need to bump it up in priority 😉.
- updated and validated interactions with some other "close" actions like...
   - "close other tabs"
   - "close tabs after"
   - "close other panes"
   - "quit" (close all windows)
- for those, I also made sure that we do an aggregate dialog instead of prompting the user once per tab/pane.
- bonus: when "close other tabs/panes" has no other tabs/panes, we don't present a dialog or handle the key binding (makes sense)

I did some decent testing but I'll be sure to test all my corner cases and update the PR body. Other than that, this is ready for review! Give it a try and let me know what you think!

**Comment by carlos-zamora** [comment]
Closing in favor of #20055 . I got to the point where 2/3 of the changes were mine, so makes sense to move it over.
